File: src/object/brdf/lambert.rs
use crate::geo;
use crate::object;

use geo::Normal3;
use geo::OrthonormalBasis;
use geo::Vec3;

use object::Color;
use object::Sampler;

use std::f32::consts::PI;

pub struct Lambert {
    strength: f32,
}

impl Lambert {
    pub fn new(strength: f32) -> Lambert {
        Lambert { strength }
    }
}

impl object::Brdf for Lambert {
    fn reflected(&self, _dir_in: Vec3, _nrm: Normal3, _dir_out: Vec3, albedo: Color) -> Color {
        return albedo * self.strength / PI;
    }

    fn transmitted(&self, _dir_in: Vec3, _nrm: Normal3, _albedo: Color) -> Color {
        return Color::ZERO;
    }

    fn lambert(&self) -> f32 {
        return self.strength;
    }

    fn sample(&self, sampler: &mut dyn Sampler, nrm: Normal3, _dir_out: Vec3) -> Vec3 {
        let sample_point = sampler.get_value2();
        let basis = OrthonormalBasis::new(nrm.into());

        let phi = 2.0 * PI * sample_point.u;
        let theta = f32::asin(sample_point.v.sqrt());

        let dir_in = basis.local_to_world(Vec3::with_spherical(phi, PI / 2.0 - theta, 1.0));

        return dir_in;
    }

    fn pdf(&self, dir_in: Vec3, nrm: Normal3, _dir_out: Vec3) -> f32 {
        let cos_theta = (dir_in * nrm).max(0.0);
        let pdf = cos_theta / PI;

        return pdf;
    }
}

File: Object/Impl/Brdf/Lambert.cpp
#define _USE_MATH_DEFINES
#include "Object/Impl/Brdf/Lambert.hpp"

#include "Math/Normal.hpp"
#include "Math/Vector.hpp"
#include "Math/OrthonormalBasis.hpp"

#include <cmath>
#include <algorithm>

namespace Object::Impl::Brdf {
    Lambert::Lambert(float strength)
    {
        mStrength = strength;
    }

    float Lambert::lambert() const
    {
        return mStrength;
    }

    Math::Color Lambert::reflected(const Math::Vector &, const Math::Normal &, const Math::Vector &, const Math::Color &albedo) const
    {
        return albedo * mStrength / (float)M_PI;
    }

    Math::Vector Lambert::sample(Math::Sampler &sampler, const Math::Normal &nrm, const Math::Vector &) const
    {
        Math::Point2D samplePoint = sampler.getValue2D();
        Math::OrthonormalBasis basis(nrm);
        float phi = 2 * M_PI * samplePoint.u();
        float theta = std::asin(std::sqrt(samplePoint.v()));

        Math::Vector dirIn = basis.localToWorld(Math::Vector::fromPolar(phi, M_PI / 2 - theta, 1));

        return dirIn;
    }

    Math::Pdf Lambert::pdf(const Math::Vector &dirIn, const Math::Normal &nrm, const Math::Vector &) const
    {
        float cosTheta = std::max(dirIn * Math::Vector(nrm), 0.0f);
        float pdf = cosTheta / M_PI;

        return pdf;
    }

    void Lambert::writeProxy(BrdfProxy &proxy) const
    {
        proxy.type = BrdfProxy::Type::Lambert;
        proxy.lambert.strength = mStrength;
    }
}